components/skus/browser/skus_sdk_service.cc

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/components/skus/browser/skus_sdk_service.h"
+
+#include <utility>
+
+#include "base/json/json_reader.h"
+#include "base/task/post_task.h"
+#include "base/threading/sequenced_task_runner_handle.h"
+#include "brave/components/skus/browser/rs/cxx/src/lib.rs.h"
+#include "brave/components/skus/browser/rs/cxx/src/shim.h"
+#include "brave/components/skus/browser/pref_names.h"
+#include "brave/components/skus/browser/skus_sdk_context_impl.h"
+#include "brave/components/skus/browser/skus_utils.h"
+#include "components/prefs/pref_service.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+
+namespace {
+
+void OnRefreshOrder(skus::RefreshOrderCallbackState* callback_state,
+                    skus::SkusResult result,
+                    rust::cxxbridge1::Str order) {
+  std::string order_str = static_cast<std::string>(order);
+  if (callback_state->cb) {
+    std::move(callback_state->cb).Run(order_str);
+  }
+  delete callback_state;
+}
+
+void OnFetchOrderCredentials(
+    skus::FetchOrderCredentialsCallbackState* callback_state,
+    skus::SkusResult result) {
+  if (callback_state->cb) {
+    std::move(callback_state->cb).Run("");
+  }
+  delete callback_state;
```

Comment:
why do we have a `delete` here? This seems like it could easily result in memory leaks. This was a unique_ptr, where did it become a raw pointer?

Reply:
unfortunately `cxx` only handles passing a rust `Fn` to C++ and doesn't handle the other way around. the way I attempted to work around this was to define a C ABI callback "trampoline" type ( whose implementation as above just calls the actual C++ callback ) which was packaged  outside of the cxx block but `UniquePtr` couldn't be passed through it: 
https://github.com/brave/brave-core/pull/10847/files/6b4899fb6d8239c4415d3471645a4f74927a8ef7#diff-47169e2c66408682410fe2d5e69b1c4fee9c405a82fbfae316b1c97313af66c1R335

as a result we're calling `into_raw` on the rust side in order to pass the callback state as a pointer back to C++, which as you've pointed out isn't a great solution.

looking at this again now, it occurs to me that we could probably just have a fixed "trampoline" defined on the C++ side ( e.g. `shim_refreshOrderCallback` ) and only pass the C++ callback to `refresh_order`, etc ( `callback_state` ) rather than passing both the C++ callback and the "trampoline". that should allow us to pass the `UniquePtr` back without issue